### DISCUSSION TITLE
Oracledb cursors management

### DISCUSSION THREAD

**Discussion by Alaeddine22** [oracle/python-oracledb#349] [Q&A]
Hello,
i'm running some tests in order to understand how oracledb is managing cursors.
I ran the following tests :

- Test1: Queries having bind variables

> import oracledb
> oracledb.defaults.stmtcachesize = 2
> conn = oracledb.connect(dsn)
> cur1 = conn.cursor()
> cur1.execute("select :1 from dual", [1]).fetchone()[0]
> cur2 = conn.cursor()
> cur2.execute("select :1 from dual", [2]).fetchone()[0]
> cur1.close()
> cur3 = conn.cursor()
> cur3.execute("select :1 from dual", [3]).fetchone()[0]
> cur2.close()
> cur4 = conn.cursor()
> cur4.execute("select :1 from dual", [3]).fetchone()[0]
> cur5 = conn.cursor()
> cur5.execute("select :1 from dual", [3]).fetchone()[0]

- Test2: Queries without bind variables 

> import oracledb
> oracledb.defaults.stmtcachesize = 2
> conn = oracledb.connect(dsn)
> cur1 = conn.cursor()
> cur1.execute("select 1 from dual").fetchone()[0]
> cur2 = conn.cursor()
> cur1.execute("select 2 from dual").fetchone()[0]
> cur1.close()
> cur3 = conn.cursor()
> cur3.execute("select 3 from dual").fetchone()[0]
> cur2.close()
> cur4 = conn.cursor()
> cur4.execute("select 3 from dual").fetchone()[0]
> cur5 = conn.cursor()
> cur5.execute("select 3 from dual").fetchone()[0]

And i noticed some strange observations listed in the attached picture
![image](https://github.com/oracle/python-oracledb/assets/41190156/5c4eaebb-fdc1-4ccb-b5e4-a0b1ad0fffca)

The query i used to check the cursors: `select * FROM  v$open_cursor where sid=[conn_sid] and cursor_type='OPEN';`

Another question: why is the cursor not released after being closed (running cur.close())? I noticed it is effectively released only when executing a query afterwards.

Thank you in advance

**Comment by cjbj** [comment]
- Thanks for using python-oracledb and starting a discussion
- What's the real problem you are trying to solve?
- Not everyone can read screenshots, so it helps the community if they are not used!
